Client/WarFare/UIItemUpgrade.cpp

```diff
@@ -0,0 +1,1289 @@
+		{
+			bTooltipRender = true;
+			spItem = GetHighlightIconItem((CN3UIIcon*) pChild);
+		}
+	}
+
+	// Display the count for items that should show a count.
+	for (int i = 0; i < MAX_ITEM_INVENTORY; i++)
+	{
+		if (m_pMyUpgradeInv[i] && ((m_pMyUpgradeInv[i]->pItemBasic->byContable == UIITEM_TYPE_COUNTABLE) ||
+			(m_pMyUpgradeInv[i]->pItemBasic->byContable == UIITEM_TYPE_COUNTABLE_SMALL)))
+		{
+			char szID[32];
+			sprintf(szID, "s_count_%d", i);
+			CN3UIString* pStr = (CN3UIString*)GetChildByID(szID);
+			if (pStr)
+			{
+				if ((GetState() == UI_STATE_ICON_MOVING) && (m_pMyUpgradeInv[i] == CN3UIWndBase::m_sSelectedIconInfo.pItemSelect))
+				{
+					pStr->SetVisible(false);
+				}
+				else
+				{
+					if (m_pMyUpgradeInv[i]->pUIIcon->IsVisible())
+					{
+						pStr->SetVisible(true);
+						pStr->SetStringAsInt(m_pMyUpgradeInv[i]->iCount);
+						pStr->Render();
+					}
+					else
+					{
+						pStr->SetVisible(false);
+					}
+				}
+			}
+		}
+		else
+		{
+			char szID[32];
+			sprintf(szID, "s_count_%d", i);
+			CN3UIString* pStr = (CN3UIString*)GetChildByID(szID);
+			if (pStr)
+				pStr->SetVisible(false);
+		}
+	}
+
+	if ((GetState() == UI_STATE_ICON_MOVING) && (CN3UIWndBase::m_sSelectedIconInfo.pItemSelect))
+		CN3UIWndBase::m_sSelectedIconInfo.pItemSelect->pUIIcon->Render();
+
+	if (spItem)
+	{
+		m_pUITooltipDlg->DisplayTooltipsEnable(ptCur.x, ptCur.y, spItem, false, false);
+	}
+}
+
+void CUIItemUpgrade::InitIconWnd(e_UIWND eWnd)
+{
+	__TABLE_UI_RESRC* pTbl = CGameBase::s_pTbl_UI.Find(CGameBase::s_pPlayer->m_InfoBase.eNation);
+
+	m_pUITooltipDlg = new CUIImageTooltipDlg();
+	m_pUITooltipDlg->Init(this);
+	m_pUITooltipDlg->LoadFromFile(pTbl->szItemInfo);
+	m_pUITooltipDlg->InitPos();
+	m_pUITooltipDlg->SetVisible(FALSE);
+
+	CN3UIWndBase::InitIconWnd(eWnd);
+
+	m_pStrMyGold = (CN3UIString*) GetChildByID("text_gold");
+	__ASSERT(m_pStrMyGold, "NULL UI Component!!");
+	if (m_pStrMyGold) m_pStrMyGold->SetString("0");
+}
+
+__IconItemSkill* CUIItemUpgrade::GetHighlightIconItem(CN3UIIcon* pUIIcon)
+{
+	int i;
+	for (i = 0; i < MAX_ITEM_UPGRADE_SLOT; i++)
+	{
+		if ((m_pMyUpgradeSLot[i] != nullptr) && (m_pMyUpgradeSLot[i]->pUIIcon == pUIIcon))
+			return m_pMyUpgradeSLot[i];
+	}
+
+	for (i = 0; i < MAX_ITEM_INVENTORY; i++)
+	{
+		if ((m_pMyUpgradeInv[i] != nullptr) && (m_pMyUpgradeInv[i]->pUIIcon == pUIIcon))
+			return m_pMyUpgradeInv[i];
+	}
+	if (m_pUpgradeItemSlot && m_pUpgradeItemSlot->pUIIcon == pUIIcon)
+		return m_pUpgradeItemSlot;
+
+	return nullptr;
+}
+
+void CUIItemUpgrade::Open()
+{
+	SetVisible(true);
+	ItemMoveFromInvToThis();
+
+	GoldUpdate();
+}
+
+void CUIItemUpgrade::GoldUpdate()
+{
+	if (m_pStrMyGold)
+	{
+		m_pStrMyGold->SetStringAsInt(CGameBase::s_pPlayer->m_InfoExt.iGold);
+	}
+}
+
+void CUIItemUpgrade::ItemMoveFromInvToThis()
+{
+	CUIInventory* pInven = CGameProcedure::s_pProcMain->m_pUIInventory;
+	if (!pInven)
+		return;
+
+
+	for (int i = 0; i < MAX_ITEM_INVENTORY; i++)
```

Comment:
This entire method seems unnecessary.
You move all items from the inventory, only to then copy them.
Why not just copy them in the first place? Then you don't have to keep the inventory up-to-date: you only need to sync it when handling the result packet (determining if it burned or succeeded).

For reference, officially it gets copied.

This just seems really easy to break.